File: agents/windows/Security/EnrollmentClient.cs
using System.Net;
using System.Net.Http;
using System.Net.Http.Json;
using System.Runtime.Versioning;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Cadence.WindowsAgent.Security;

/// <summary>
/// Client for the brain's un-authenticated device-enrollment intake
/// (<c>POST /device/enroll</c>, <c>cadence/brain/app.py</c> → <c>cadence/devices/enrollment.py</c>).
///
/// <para>
/// This is the deliberate chicken-and-egg exception to the mTLS gate: an un-enrolled device has no
/// client certificate yet, so this endpoint is NOT behind mTLS and this client does NOT present a
/// client cert. It still MUST validate the brain's <b>server</b> certificate against the pinned
/// brain CA — configure that on the injected <see cref="HttpClient"/>'s handler (e.g. a
/// <c>SocketsHttpHandler</c> with <c>SslOptions.RemoteCertificateValidationCallback</c> pinned to
/// the CA), the same CA that <c>ca_pem</c> pins for the post-enrollment mTLS transport.
/// </para>
///
/// <para>
/// Proof of possession is required (a bare public key is rejected). We send a DPoP-style proof:
/// <c>dpop_proof = base64( DER ECDSA-P256/SHA-256 signature by the device key over the device's own
/// DER SPKI )</c>, produced by <see cref="TpmClientAuthSigner.SignToDer"/> over
/// <see cref="TpmClientAuthSigner.PublicKeySpkiDer"/>. The brain verifies it with
/// <c>public_key.verify(sig, spki_der, ec.ECDSA(SHA256))</c>.
/// </para>
/// </summary>
[SupportedOSPlatform("windows")]
internal sealed class EnrollmentClient
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    private readonly HttpClient _http;

    /// <param name="http">
    /// An <see cref="HttpClient"/> whose <c>BaseAddress</c> is the brain base URL and whose handler
    /// pins the brain CA for server-cert validation (see the type remarks). Ownership stays with the
    /// caller / DI container.
    /// </param>
    internal EnrollmentClient(HttpClient http)
    {
        _http = http ?? throw new ArgumentNullException(nameof(http));
    }

    /// <summary>
    /// Build the enrollment payload from <paramref name="signer"/> and POST it to
    /// <c>/device/enroll</c>. Returns the parsed result; a successful enrollment lands the device in
    /// the <c>pending</c> state (see <see cref="EnrollmentResult.IsPending"/>) awaiting an operator
    /// accept — no certificate is issued here.
    /// </summary>
    /// <exception cref="HttpRequestException">The request could not be completed (network/TLS).</exception>
    internal async Task<EnrollmentResult> EnrollAsync(
        string deviceName, TpmClientAuthSigner signer, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(deviceName);
        ArgumentNullException.ThrowIfNull(signer);

        byte[] spkiDer = signer.PublicKeySpkiDer();
        string dpopProof = Convert.ToBase64String(signer.SignToDer(spkiDer));

        var payload = new EnrollmentRequestDto
        {
            Name = deviceName,
            PublicKey = signer.PublicKeySpkiPem(),
            KeyProvenance = signer.KeyProvenance(),
            DpopProof = dpopProof,
        };

        using HttpResponseMessage response = await _http
            .PostAsJsonAsync("/device/enroll", payload, JsonOptions, cancellationToken)
            .ConfigureAwait(false);

        // The route returns 201 on a fresh/idempotent pending row, 422 on a bad payload/PoP, and
        // 429 when the anonymous pending-device cap is hit. All three carry a JSON body.
        EnrollmentResponseDto? body = await response.Content
            .ReadFromJsonAsync<EnrollmentResponseDto>(JsonOptions, cancellationToken)
            .ConfigureAwait(false);

        return EnrollmentResult.From(response.StatusCode, body);
    }

    /// <summary>Wire payload for <c>EnrollmentRequest</c> (extra fields forbidden by the brain).</summary>
    private sealed class EnrollmentRequestDto
    {
        [JsonPropertyName("name")]
        public required string Name { get; init; }

        [JsonPropertyName("public_key")]
        public required string PublicKey { get; init; }

        [JsonPropertyName("key_provenance")]
        public required string KeyProvenance { get; init; }

        // Omit-on-null so we never send `csr`/`attestation`; the brain's model is `extra: forbid`
        // and these optionals default to None — sending only what we set keeps the payload minimal.
        [JsonPropertyName("dpop_proof")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string? DpopProof { get; init; }
    }

    /// <summary>
    /// Union of the route's success body (<c>device_id</c>, <c>status</c>, <c>public_key_fingerprint</c>)
    /// and its error body (<c>error</c>, <c>detail</c>). Unset fields stay null.
    /// </summary>
    internal sealed class EnrollmentResponseDto
    {
        [JsonPropertyName("device_id")]
        public string? DeviceId { get; init; }

        [JsonPropertyName("status")]
        public string? Status { get; init; }

        [JsonPropertyName("public_key_fingerprint")]
        public string? PublicKeyFingerprint { get; init; }

        [JsonPropertyName("error")]
        public string? Error { get; init; }

        [JsonPropertyName("detail")]
        public string? Detail { get; init; }
    }

    /// <summary>Outcome of an enrollment attempt.</summary>
    internal sealed record EnrollmentResult(
        HttpStatusCode StatusCode,
        string? DeviceId,
        string? Status,
        string? PublicKeyFingerprint,
        string? Error,
        string? Detail)
    {
        /// <summary>True when the device was recorded and is awaiting an operator accept.</summary>
        public bool IsPending =>
            StatusCode == HttpStatusCode.Created &&
            string.Equals(Status, "pending", StringComparison.Ordinal);

        internal static EnrollmentResult From(HttpStatusCode status, EnrollmentResponseDto? body) => new(
            status,
            body?.DeviceId,
            body?.Status,
            body?.PublicKeyFingerprint,
            body?.Error,
            body?.Detail);
    }
}

File: agents/android/app/src/main/java/com/cadence/agent/enrollment/EnrollmentClient.kt
package com.cadence.agent.enrollment

import com.cadence.agent.core.StrongBoxSigner
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64

/**
 * Client for the brain's un-authenticated device-enrollment intake
 * (`POST /device/enroll`, `cadence/devices/enrollment.py`) — the one endpoint reachable
 * **without** a client cert, because an un-enrolled device has none yet (the mTLS
 * chicken-and-egg boundary).
 *
 * ## Flow (Trust-On-First-Use + explicit human accept)
 * 1. This client builds an [EnrollmentRequest] from a [StrongBoxSigner]: the key's SPKI
 *    PEM as `public_key`, `key_provenance = "hardware"` (StrongBox/TEE both map here), and
 *    a `dpop_proof` = base64 of the signer's DER ECDSA signature over its **own DER SPKI**
 *    (proof-of-possession — anti-squatting; the brain verifies it against `public_key`).
 * 2. It POSTs the request. The brain records a single **`pending`** device row and returns
 *    `201` with `{status: "pending", device_id, public_key_fingerprint, ...}`. No cert is
 *    issued and no trust is conferred by enrolling — [EnrollmentResult.Pending].
 * 3. **Out of band**, a human operator reviews the pending device and runs
 *    `cadence device accept <device_id>` (the operator CLI + cert issuance are B2). The
 *    issued client certificate chain (leaf first) and the brain's CA pin are delivered
 *    back to the device out-of-band.
 * 4. The agent then brings up the mTLS transport with the StrongBox key still behind the
 *    seam: `CoreBridge.initWithSigner(configJson, strongBoxSigner)`, where `configJson`
 *    carries the issued `cert_chain_pem` + `ca_pem` (see [CoreBridge] and
 *    `agents/core/src/ffi.rs::SignerCoreConfig`).
 *
 * ## Transport trust
 * Enrollment runs before the device trusts the brain's CA via the core, so the brain's
 * server cert must still be validated here. Pass the brain CA PEM as [caPem] to pin it;
 * with `null`, the platform trust store is used (only appropriate when the brain presents
 * a publicly-trusted cert). Pinning is done by the caller-supplied [connectionFactory].
 *
 * **Device/network validation pending:** exercised on-device against a live brain; the
 * request/response shaping and PoP construction are compiled and doc-checked here.
 *
 * @param baseUrl brain base URL, e.g. `https://brain.example:8443`.
 * @param caPem optional brain CA PEM to pin (documentation of intent; wiring the pinned
 *   `SSLSocketFactory` is the caller's responsibility via [connectionFactory]).
 * @param connectionFactory opens a connection for a URL; override to inject a pinned
 *   `HttpsURLConnection` (default uses [URL.openConnection]).
 */
class EnrollmentClient(
    private val baseUrl: String,
    @Suppress("unused") private val caPem: String? = null,
    private val connectionFactory: (URL) -> HttpURLConnection = { url ->
        url.openConnection() as HttpURLConnection
    },
) {
    @OptIn(ExperimentalSerializationApi::class) // explicitNulls
    private val json = Json {
        // extra=forbid on the server: send exactly the fields we set, drop nulls
        // (csr/attestation stay absent) rather than emitting `null`s that would 422.
        explicitNulls = false
        encodeDefaults = true
        ignoreUnknownKeys = true
    }

    /**
     * Build and POST the enrollment request for [signer], labelling the device [deviceName].
     * An optional [attestationRef] (a NAS pointer to the uploaded Key Attestation chain,
     * NOT the raw blob) is carried through for the operator's B2 review.
     *
     * @throws EnrollmentException on a transport failure or a non-2xx/non-422/429 response.
     */
    fun enroll(
        signer: StrongBoxSigner,
        deviceName: String,
        attestationRef: String? = null,
    ): EnrollmentResult {
        val request = EnrollmentRequest(
            name = deviceName,
            publicKey = signer.publicKeySpkiPem(),
            // StrongBox and TEE are both hardware-resident; the brain's key_provenance
            // vocabulary is hardware|software|unknown. The fine-grained strongbox/tee
            // value lives in signer.keyProvenance() for local telemetry.
            keyProvenance = "hardware",
            dpopProof = Base64.getEncoder().encodeToString(signer.signSpki()),
            attestation = attestationRef,
        )
        return post(request)
    }

    private fun post(request: EnrollmentRequest): EnrollmentResult {
        val url = URL(baseUrl.trimEnd('/') + "/device/enroll")
        val body = json.encodeToString(EnrollmentRequest.serializer(), request)
            .toByteArray(Charsets.UTF_8)
        val conn = connectionFactory(url).apply {
            requestMethod = "POST"
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Accept", "application/json")
            connectTimeout = CONNECT_TIMEOUT_MS
            readTimeout = READ_TIMEOUT_MS
        }
        try {
            conn.outputStream.use { it.write(body) }
            val code = conn.responseCode
            val stream = if (code in 200..299) conn.inputStream else conn.errorStream
            val text = stream?.readBytes()?.toString(Charsets.UTF_8) ?: ""
            return parse(code, text)
        } catch (e: IOException) {
            throw EnrollmentException("enrollment POST failed: ${e.message}", e)
        } finally {
            conn.disconnect()
        }
    }

    private fun parse(code: Int, text: String): EnrollmentResult {
        return when (code) {
            201, 200 -> {
                val resp = json.decodeFromString(EnrollmentResponse.serializer(), text)
                // The brain returns pending; never trusted/accepted on enroll.
                EnrollmentResult.Pending(
                    deviceId = resp.deviceId,
                    status = resp.status ?: "pending",
                    fingerprint = resp.publicKeyFingerprint,
                )
            }
            422 -> EnrollmentResult.Rejected(detail = errorDetail(text) ?: "invalid enrollment")
            429 -> EnrollmentResult.CapacityExceeded(detail = errorDetail(text) ?: "pending capacity")
            else -> throw EnrollmentException("unexpected enrollment status $code: $text")
        }
    }

    private fun errorDetail(text: String): String? =
        runCatching { json.decodeFromString(EnrollmentResponse.serializer(), text).detail }
            .getOrNull()

    companion object {
        private const val CONNECT_TIMEOUT_MS = 15_000
        private const val READ_TIMEOUT_MS = 15_000
    }
}

/**
 * The enrollment payload, field-for-field aligned with
 * `cadence.devices.enrollment.EnrollmentRequest` (`model_config = extra: forbid`). Only
 * `name`, `public_key`, `key_provenance`, and `dpop_proof` are always sent; `csr` is
 * never sent (we prove possession via DPoP, not a CSR) and `attestation` is sent only when
 * a NAS reference is supplied. Nulls are dropped by the client's `explicitNulls = false`.
 */
@Serializable
data class EnrollmentRequest(
    val name: String,
    @SerialName("public_key") val publicKey: String,
    @SerialName("key_provenance") val keyProvenance: String,
    @SerialName("dpop_proof") val dpopProof: String,
    val attestation: String? = null,
)

/** The brain's `/device/enroll` response body (`cadence/brain/app.py`). */
@Serializable
data class EnrollmentResponse(
    @SerialName("device_id") val deviceId: String? = null,
    val status: String? = null,
    @SerialName("public_key_fingerprint") val publicKeyFingerprint: String? = null,
    val accepted: Boolean = false,
    val trusted: Boolean = false,
    // Present on 422/429 error bodies.
    val error: String? = null,
    val detail: String? = null,
)

/** Outcome of an enrollment attempt. */
sealed class EnrollmentResult {
    /**
     * The brain recorded a `pending` device. No cert yet — an operator must run
     * `cadence device accept` and deliver the client cert + CA pin out-of-band, after
     * which the agent calls `CoreBridge.initWithSigner(...)`.
     */
    data class Pending(
        val deviceId: String?,
        val status: String,
        val fingerprint: String?,
    ) : EnrollmentResult()

    /** The brain rejected the payload (bad key / bad PoP) — HTTP 422. */
    data class Rejected(val detail: String) : EnrollmentResult()

    /** The anonymous pending-device cap is full — HTTP 429; retry later. */
    data class CapacityExceeded(val detail: String) : EnrollmentResult()
}

/** A transport-level or unexpected-status enrollment failure. */
class EnrollmentException(message: String, cause: Throwable? = null) :
    RuntimeException(message, cause)
